README.md
# edimax-smartplug

[![Greenkeeper badge](https://badges.greenkeeper.io/mwittig/edimax-smartplug.svg)](https://greenkeeper.io/) [![Build Status](https://travis-ci.org/mwittig/edimax-smartplug.svg)](https://travis-ci.org/mwittig/edimax-smartplug)

Node module to communicate with Edimax Smart Plugs. The library utilizes Bluebird 
promises - https://github.com/petkaantonov/bluebird. Device requests will be executed sequentially. 
This is useful as the Smart Plug REST service does not properly chain incoming requests. It provides the 
following features:

* Device Discovery - Smart Plugs can be automatically discovered on a given network 
  (see section "Device Discovery" below for details)
* Switching - Apart from switching Smart Plugs on and off, it is possible to read the current switch state 
* Schedule - It is not possible to set the schedule for the Smart Plugs, but the current schedule is readable 
* Usage Metering - For 'SP2101W' Smart Plugs it is possible to access the current, accumulated, and historical
  energy usage metering data
* Support for digest authentication required with firmware versions SP-2101W v2.08 and SP-1101W v2.04

| Important Notice |
|------------------|
The recent firmware versions published for EdiSmart/Amazon Alexa Integration 
are currently *not supported* by edimax-smartplug:
* SP2101W: v2.09 and higher
* SP1101W: v1.05 and higher

Moreover, please note, the SP-2101W V2 plug is currently not supported. 
 [Issue #14](https://github.com/mwittig/edimax-smartplug/issues/14) contains a link thread with 
 untested and unsupported downgrade instructions. Use at your own risk!

If you have installed the new version and wish to downgrade you can use the 
[downgrade guide](https://github.com/mwittig/edimax-smartplug/blob/master/Downgrade.md) provided 
as part of the project.

## Usage Examples

    var smartplug = require('./index');
    var options = {
        timeout: 10000,
        name:'edimax',
        host:'192.168.178.65',
        username: 'admin',
        password: '1234'
    };
    
    
    smartplug.getDeviceInfo(options).then(function (info) {
        console.log(info);
    }).catch(function(e) {console.log("Request failed: ", e)});
    
    smartplug.getSchedule(options).then(function (schedule) {
        console.log(schedule);
    }).catch(function(e) {console.log("Request failed: ", e)});
    
    // set switch ON
    smartplug.setSwitchState(true, options).catch(function(e) {console.log("Request failed: ", e)});
    
    smartplug.getSwitchPower(options).then(function (power) {
        console.log("Current switch power", power, "Watts");
    }).catch(function(e) {console.log("Request failed: ", e)});
    
    smartplug.getSwitchEnergy(options).then(function (energy) {
        console.log("getSwitchEnergy result:", energy);
    }).catch(function(e) {console.log("Request failed: ", e)});
    
    smartplug.getStatusValues(true, options).then(function (all) {
        console.log("getStatusValues result:", all);
    }).catch(function(e) {console.log("Request failed: ", e)});
    
    // set switch OFF
    smartplug.setSwitchState(false, options).catch(function(e) {console.log("Request failed: ", e)});
    
    // get switch status
    smartplug.getSwitchState(options).then(function (state) {
        console.log("Switch status is", state?"ON":"OFF");
    }).catch(function(e) {console.log("Request failed: ", e)});

    // get schedule status
    smartplug.getScheduleState(options).then(function (state) {
        console.log("Schedule status is", state?"ON":"OFF");
    }).catch(function(e) {console.log("Request failed: ", e)});
    
    // get the daily history of power measured consumption for the given date range
    smartplug.getHistory('DAY', '20160825', '20160830', options).then(function (results) {
        console.log("getHistory result", results);
    }).catch(function(e) {console.log("Request failed: ", e)});
    
    // discover devices
    smartplug.discoverDevices({
        timeout: 3000,
        address: "192.168.178.255"
    }).then(function (results) {
        console.log("Discovery Result:", results);
    }).catch(function(e) {console.log("Request failed: ", e)});
    
## Device Discovery

The device discovery implementation is based on the findings summarized in a 
[blog post](http://blog.guntram.de/?p=45) (thanks, Guntram). As the discovery 
mechanism may also be used for other Edimax products, e.g. IP cameras, you should filter
by model name to make sure the found device is a smart plug ('SP1101W' and 'SP2101W'). 
The method `discoverDevices()` accepts the following options:

| Property  | Default           | Type    | Description                                 |
|:----------|:------------------|:--------|:--------------------------------------------|
| address   | "255.255.255.255" | String  | The broadcast address                       |
| timeout   | 3000              | Integer | The timeout in milliseconds for discovery   |

Note: Using the global broadcast address on Windows may yield unexpected results. On Windows, 
global broadcast packets will only be routed via the first network adapter which may cause problems
with multi-homed setups and virtual network adapters. If you want to use a broadcast address 
though, use a network-specific address, e.g. for `192.168.0.1/24` use `192.168.0.255`.

A discovery request returns a promise which yields an array with the discovery items on completion. An item is 
an object with the following properties:

| Property      | Type    | Example          | Description                    |
|:--------------|:--------|:-----------------|:-------------------------------|
| manufacturer  | String  | 'EDIMAX'         | The manufacturer of the device |
| model         | String  | 'SP2101W'        | The model name                 |
| version       | String  | '2.03'           | The firmware version           |
| displayName   | String  | 'edimax'         | The assigned device name       |
| addr          | String  | '192.168.178.65' | The IP address of the device   |
| dstAddr       | String  | '192.168.178.1'  | The IP address of the gateway  |
    
## TODO

* More Documentation
* Make request chaining optional. In some cases, request/response interleaving does not matter
* Add tests

## History

See [Release History](https://github.com/mwittig/edimax-smartplug/blob/master/HISTORY.md).

## License 

Copyright (c) 2015-2019, Marcus Wittig and contributors. All rights reserved.

[MIT License](https://github.com/mwittig/edimax-smartplug/blob/master/LICENSE)


HISTORY.md
# Release History

* 20190328, V0.0.24
    * Added node v10 build to travis build regime
    * Added note about unsupported SP-2101W V2, issue #14
    * Dependency updates
    * Revised Copyright Notice
    
* 20170218, V0.0.23
    * Dependency updates
    * Revised Firmware Downgrade Guide and Copyright Notice
    
* 20171218, V0.0.22
    * Dependency updates
    * Updated Travis build descriptor
    * Revised README, added incompatibility notice
    * Added Firmware Downgrade Guide
    
* 20171204, V0.0.21
    * Dependency updates

* 20170514, V0.0.20
    * Fix: Syntax Error. Misplaced semicolon result in globals
    
* 20170404, V0.0.19
    * Feature: Added a private HTTP connection agent as using the global agent may have undesirable side effects
      in application with multiple connection targets
    * Fix: Use util.inspect instead of JSON.stringify to avoid circular dependencies error if debugging is enabled
    * Updated and revised dependencies. Now using Object.assign instead of lodash assign

* 20170029, V0.0.18
    * Feature: Added support for digest authentication required with firmware versions SP-2101W v2.08 and SP-1101W v2.04
    * Integrated modified version of http-digest-auth with fallback to basic authentication to
      support older firmware versions

* 20170118, V0.0.17
    * Feature: Added method to get the current schedule state (contributed by @mplessing)
    * Dependency updates
    
* 20161207, V0.0.16
    * Dependency updates
    * Revised README
    * Updated Travis build descriptor
    
* 20160831, V0.0.15
    * Feature: Added method to access history data (contributed by @mattjgalloway)
    * Updated README and demo.js
    * Dependency updates
    
* 20160503, V0.0.14
    * Bug fix: Close discovery socket on error
    * Improved error handling
    
* 20160503, V0.0.13
    * Fix: ensure discovery service will timeout if no results received

* 20160425, V0.0.12
    * Added device discovery service
    * Dependency updates
    * Replaced lodash with lightweight lodash method modules
    * Added release history file
    * Updated README
    
* 20160305, V0.0.11
    * Dependency updates
    * Replaced deprecated use of Promise.settle()
    * Minor changes of demo.js

* 20151231, V0.0.10
    * Added missing commit from previous releas
    
* 20151231, V0.0.9
    * Bug fix. Always set SmartPlug id XML-Attribute to "edimax" rather than the name assigned to the plug.
    * Dependency update
    * Added basic Travis builds
  
* 20150511, V0.0.8
    * Enforce a default timeout of 20000 msecs to cleanup if client is connected but server does not send a response.

* 20150508, V0.0.7
    * Fixed request error handling. Request must be aborted if error has occurred
    * Added option to set a request timeout
    
* 20150416, V0.0.6
    * Fixed bug in request processing which caused a TypeError if withMetering was set to false

* 20150413, V0.0.5
    * Added getStatusValues() method to provide bulk updates optionally including metering values
    
* 20150413, V0.0.4
    * Improved error handling (Unauthorized case, in particular)
    * Handle chunked response data
    * Added basic implementation to obtain schedule info
    * Updated README
    
* 20150413, V0.0.3
    * Added function to get Energy & Power attributes and to read device information (vendor, model, firmware version)
    
* 20150412, V0.0.2
    * Corrected package descriptor
    
* 20150412, V0.0.1
    * Initial Version


index.js
var util = require('util'),
    http = require('http-digest-client'),
    _isUndefined = require('lodash.isundefined'),
    _parseInt = require('lodash.parseint'),
    xpath = require('xpath'),
    xmldom = require('xmldom'),
    Promise = require('bluebird'),
    dgram = require('dgram'),
    commandTemplateString = '<?xml version="1.0" encoding="UTF8"?><SMARTPLUG id="edimax"><CMD id="%s">%s</CMD></SMARTPLUG>',
    lastRequest = Promise.resolve(),
    debug = process.env.hasOwnProperty('EDIMAX_DEBUG') ? consoleDebug : function () {
    },
    nodeHttp = require('http')
var agent = new nodeHttp.Agent({ keepAlive: false });

require('es6-object-assign/auto');

//
// Private Helper Functions
//

function createCommandString(deviceName, command, commandXml) {
    return util.format(commandTemplateString, command, commandXml);
}

function consoleDebug() {
    console.log.apply(this, arguments)
}

function settlePromise(aPromise) {
    return aPromise.reflect();
}

function assignDefaultCommandOptions(options) {
    return Object.assign({}, {
        name: 'edimax'
    }, options)
}

function postRequest(command, options) {
    var requestOptions = Object.assign({}, {
            timeout: 20000,
            port: 10000,
            path: 'smartplug.cgi',
            method: 'POST',
            headers: {
                'Content-Type': 'application/xml',
                'Content-Length': command.length
            },
            username: 'admin',
            password: '1234',
            agent: agent
        }, options),
        timeoutOccurred = false;

    debug('REQUEST OPTIONS: ' + util.inspect(requestOptions));
    debug('REQUEST: ' + command);

    return new Promise(function (resolve, reject) {
        var data = "";
        var client = http(requestOptions.username, requestOptions.password);
        var postReq = client.request(requestOptions, function (response) {
            debug('STATUS: ' + response.statusCode);
            debug('HEADERS: ' + util.inspect(response.headers));

            var error;
            if (response.statusCode >= 300) {
                if (response.statusCode === 401) {
                    error = new Error("Unauthorized: check username/password");
                }
                else {
                    error = new Error("Request failed. HTTP Status Code: " + response.statusCode);
                }
                debug('ERROR:' + 'Host ' + requestOptions.host + ' ' + error);
                return reject(error);
            }
            var contentLength = response.headers['content-length'];
            if (_isUndefined(contentLength) || _parseInt(contentLength) === 0) {
                error = new Error("No such device: check name");
                debug('ERROR:' + 'Host ' + requestOptions.host + ' ' + error);
                return reject(error);
            }

            response.setEncoding('utf8');
            response.on('data', function (result) {
                debug("DATA CHUNK", result);
                data += result;
            });
            response.on('end', function () {
                debug("END");
                var doc = new xmldom.DOMParser().parseFromString(data, 'text/xml');
                return resolve(doc);
            });
        });
        postReq.on('error', function (error) {
            if (timeoutOccurred) {
                error = new Error("Request timeout occurred - request aborted");
            }
            debug('ERROR:' + 'Host ' + requestOptions.host + ' ' + error);
            postReq.abort();
            return reject(error);
        });
        postReq.on('timeout', function () {
            timeoutOccurred = true;
            postReq.abort();
        });
        postReq.setTimeout(requestOptions.timeout);
        postReq.write(command);
        postReq.end();
    });
}

function bufIndexOf(buffer, byteVal, start, end) {
    var pos = start;
    for (; pos < end; pos++) {
        if (buffer[pos] === byteVal) {
            break;
        }
    }
    return pos;
}

function bytesToIpAddress(bytes) {
    if (bytes.length === 4) {
        return "" + bytes[0] + "." + bytes[1] + "." + bytes[2] + "." + bytes[3];
    }
    else {
        throw new Error("Buffer does not contain valid IPv4 address")
    }
}

function toHexString(d) {
    return ("0"+(Number(d).toString(16))).slice(-2).toUpperCase()
}

function bytesToMacAddress(bytes) {
    if (bytes.length === 6) {
        return "" + toHexString(bytes[0]) + ":" + toHexString(bytes[1]) + ":" + toHexString(bytes[2])
            + ":" + toHexString(bytes[3]) + ":" + toHexString(bytes[4]) + ":" + toHexString(bytes[5]);
    }
    else {
        throw new Error("Buffer does not contain valid MAC address")
    }
}

function numberFromCharCode(charCode) {
    if (charCode >= '0'.charCodeAt(0) && charCode <= '9'.charCodeAt(0)) {
        return charCode - '0'.charCodeAt(0)
    }
    if (charCode >= 'a'.charCodeAt(0) && charCode <= 'z'.charCodeAt(0)) {
        return charCode - 'a'.charCodeAt(0) + 10
    }
    if (charCode >= 'A'.charCodeAt(0) && charCode <= 'Z'.charCodeAt(0)) {
        return charCode - 'A'.charCodeAt(0) + 36
    }
    if (charCode == '+'.charCodeAt(0)) {
        return 62
    }
    if (charCode == '/'.charCodeAt(0)) {
        return 63
    }
    throw new Error("Invalid character code: " + charCode + "(" + String.fromCharCode(charCode) + ")");
}

function decodeHistoryValue(value) {
    var result = 0;

    if (value == '=') {
        return 0;
    }

    for (var i = 0; i < value.length; i++) {
        var thisChar = value.charCodeAt(value.length - i - 1);
        var thisNumber = numberFromCharCode(thisChar);
        result += thisNumber * (Math.pow(64, i))
    }

    return result / 1000;
}

function decodeScheduleActive(schedule) {
    var date = new Date();
    var full_min = date.getMinutes() + date.getHours() * 60;
    var start_min = 0;
    var end_min = 0;
    var day_of_week = date.getDay();
    var daily_schedule = schedule[day_of_week];
    var sched_items = daily_schedule.split('-');
    var result = false;
    for (var i = 0; i < sched_items.length && result !== true; i++) {
       	if (sched_items[i].endsWith('1')) {
            start_min = numberFromCharCode(sched_items[i].charCodeAt(0)) * 60 + numberFromCharCode(sched_items[i].charCodeAt(1));
            end_min = numberFromCharCode(sched_items[i].charCodeAt(2)) * 60 + numberFromCharCode(sched_items[i].charCodeAt(3));
            result = (full_min >= start_min && full_min <= end_min)
        }
    }
    return result;
}

//
// Public Functions
//

module.exports.getScheduleState = function (options) {
    return this.getSchedule(options).then(function (schedule) {
	      return Promise.resolve(decodeScheduleActive(schedule));
    });
};

module.exports.getSwitchState = function (options) {
    var commandOptions = assignDefaultCommandOptions(options),
        commandString = createCommandString(commandOptions.name, "get", "<Device.System.Power.State/>");

    return lastRequest = settlePromise(lastRequest).then(function () {
        return postRequest(commandString, commandOptions).then(function (responseDom) {
            return Promise.resolve(
                /^ON$/.test(xpath.select("//Device.System.Power.State/text()", responseDom).toString())); // true if on
        })
    });
};

module.exports.setSwitchState = function (state, options) {
    var commandOptions = assignDefaultCommandOptions(options),
        command = util.format("<Device.System.Power.State>%s</Device.System.Power.State>", state ? "ON" : "OFF"),
        commandString = createCommandString(commandOptions.name, "setup", command);

    return lastRequest = settlePromise(lastRequest).then(function () {
        return postRequest(commandString, commandOptions).then(function () {
            return Promise.resolve();
        })
    });
};

module.exports.getSwitchPower = function (options) {
    var commandOptions = assignDefaultCommandOptions(options),
        commandString = createCommandString(commandOptions.name, "get",
            "<NOW_POWER><Device.System.Power.NowPower/></NOW_POWER>");

    return lastRequest = settlePromise(lastRequest).then(function () {
        return postRequest(commandString, commandOptions).then(function (responseDom) {
            return Promise.resolve(
                parseFloat(xpath.select("//Device.System.Power.NowPower/text()", responseDom).toString())
            );
        })
    });
};

module.exports.getSwitchEnergy = function (options) {
    var commandOptions = assignDefaultCommandOptions(options),
        commandString = createCommandString(commandOptions.name, "get",
            "<NOW_POWER><Device.System.Power.NowEnergy.Day/><Device.System.Power.NowEnergy.Week/><Device.System.Power.NowEnergy.Month/></NOW_POWER>");

    return lastRequest = settlePromise(lastRequest).then(function () {
        return postRequest(commandString, commandOptions).then(function (responseDom) {
            return Promise.resolve({
                day: parseFloat(xpath.select("//Device.System.Power.NowEnergy.Day/text()", responseDom).toString()),
                week: parseFloat(xpath.select("//Device.System.Power.NowEnergy.Week/text()", responseDom).toString()),
                month: parseFloat(xpath.select("//Device.System.Power.NowEnergy.Month/text()", responseDom).toString())
            });
        })
    });
};

module.exports.getStatusValues = function (withMetering, options) {
    var commandOptions = assignDefaultCommandOptions(options),
        command = withMetering ? "<Device.System.Power.State/><SCHEDULE/><NOW_POWER/>" : "<Device.System.Power.State/><SCHEDULE/>",
        commandString = createCommandString(commandOptions.name, "get", command);

    return lastRequest = settlePromise(lastRequest).then(function () {
        return postRequest(commandString, commandOptions).then(function (responseDom) {
            var result = {
                state: /^ON$/.test(xpath.select("//Device.System.Power.State/text()", responseDom).toString()),
                scheduleState: false,
                nowPower: 0,
                nowCurrent: 0,
                day: 0,
                week: 0,
                month: 0
            };
            var schedule = {};
            for (var x = 0; x <= 6; ++x) {
                schedule[x] = xpath.select("//Device.System.Power.Schedule." + x + ".List/text()", responseDom).toString()
            }
            result = Object.assign(result, {
                scheduleState: decodeScheduleActive(schedule)
            });
            if (withMetering) {
                var toggleTime = xpath.select("//Device.System.Power.LastToggleTime/text()", responseDom).toString(),
                    date = (toggleTime.length === 14) ?
                        new Date(
                            toggleTime.substring(0, 4),
                            parseInt(toggleTime.substring(4, 6)) - 1,
                            toggleTime.substring(6, 8),
                            toggleTime.substring(8, 10),
                            toggleTime.substring(10, 12),
                            toggleTime.substring(12, 14)
                        ) : new Date();
                result = Object.assign(result, {
                    lastToggleTime: date,
                    nowPower: parseFloat(xpath.select("//Device.System.Power.NowPower/text()", responseDom).toString()),
                    nowCurrent: parseFloat(xpath.select("//Device.System.Power.NowCurrent/text()", responseDom).toString()),
                    day: parseFloat(xpath.select("//Device.System.Power.NowEnergy.Day/text()", responseDom).toString()),
                    week: parseFloat(xpath.select("//Device.System.Power.NowEnergy.Week/text()", responseDom).toString()),
                    month: parseFloat(xpath.select("//Device.System.Power.NowEnergy.Month/text()", responseDom).toString())
                })
            }
            return Promise.resolve(result);
        })
    });
};

module.exports.getSchedule = function (options) {
    var commandOptions = assignDefaultCommandOptions(options),
        commandString = createCommandString(commandOptions.name, "get",
            "<SCHEDULE></SCHEDULE>");

    return lastRequest = settlePromise(lastRequest).then(function () {
        return postRequest(commandString, commandOptions).then(function (responseDom) {
            var result = {};
            for (var x = 0; x <= 6; ++x) {
                result[x] = xpath.select("//Device.System.Power.Schedule." + x + ".List/text()", responseDom).toString()
            }
            return Promise.resolve(result);
        })
    });
};

module.exports.getDeviceInfo = function (options) {
    var commandOptions = assignDefaultCommandOptions(options),
        commandString = createCommandString(commandOptions.name, "get",
            "<SYSTEM_INFO><Run.Cus/><Run.Model/><Run.FW.Version/><Run.LAN.Client.MAC.Address/></SYSTEM_INFO>");

    return lastRequest = settlePromise(lastRequest).then(function () {
        return postRequest(commandString, commandOptions).then(function (responseDom) {
            return Promise.resolve({
                vendor: xpath.select("//Run.Cus/text()", responseDom).toString(),
                model: xpath.select("//Run.Model/text()", responseDom).toString(),
                fwVersion: xpath.select("//Run.FW.Version/text()", responseDom).toString(),
                mac: xpath.select("//Run.LAN.Client.MAC.Address/text()", responseDom).toString()
            });
        })
    });
};

module.exports.getHistory = function (unit, startDate, endDate, options) {
    var commandOptions = assignDefaultCommandOptions(options),
        commandString = createCommandString(commandOptions.name, "get",
            util.format("<POWER_HISTORY><Device.System.Power.History.Energy unit=\"%s\" date=\"%s-%s\"/></POWER_HISTORY>", unit, startDate, endDate));

    return lastRequest = settlePromise(lastRequest).then(function () {
        return postRequest(commandString, commandOptions).then(function (responseDom) {
            var results = xpath.select("//Device.System.Power.History.Energy/text()", responseDom).toString();
            var resultsArray = results.split('-');
            var decodedResultsArray = [];
            for (i = 0; i < resultsArray.length; i++) {
                decodedResultsArray.push(decodeHistoryValue(resultsArray[i]));
            }
            return Promise.resolve(decodedResultsArray);
        })
    });
};

module.exports.discoverDevices = function (options) {
    var options = options || {};
    var port = options.port || 20560;
    var host = options.address || "255.255.255.255";
    var timeout = options.timeout || "3000";
    var discoveryMessage = Buffer([
        0xFF, 0xFF, 0xFF, 0xFF,
        0xFF, 0xFF, 0x45, 0x44,
        0x49, 0x4d, 0x41, 0x58,
        0x00, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x00, 0xA1,
        0xFF, 0x5E
    ]);
    var timeoutId = null;
    var discoResults = [];

    return new Promise(function (resolve, reject) {
        var discoverer = dgram.createSocket('udp4');
        discoverer.bind();

        discoverer.on('listening', function () {

            try {
                discoverer.setBroadcast(true);

                if (typeof host !== "string") {
                    throw new TypeError("invalid arguments: IP address must be a string");
                }
                discoverer.send(discoveryMessage, 0, discoveryMessage.length, port, host, function(error, bytes) {
                    if (error) {
                        discoverer.emit('error', error);
                    }
                    else {
                        debug('UDP message sent to ' + host +':'+ port);

                        timeoutId = setTimeout(function() {
                            try {
                                discoverer.close();
                            } catch (ex) {/*ignore*/}
                            resolve(discoResults);
                        }, timeout)
                    }
                });
            }
            catch (e) {
                discoverer.emit('error', e);
            }
        });

        discoverer.on('message', function (message, remote) {
            discoResults.push({
                mac: bytesToMacAddress(message.slice(0, 6)),
                manufacturer: message.toString('ascii', 6, bufIndexOf(message, 0x00, 6, 18)),
                model: message.toString('ascii', 22, bufIndexOf(message, 0x00, 22, 36)),
                version: message.toString('ascii', 36, bufIndexOf(message, 0x00, 36, 44)),
                displayName: message.toString('ascii', 44, bufIndexOf(message, 0x00, 44, 172)),
                port: message.readInt16LE(172),
                addr: bytesToIpAddress(message.slice(174, 178)),
                dstAddr: bytesToIpAddress(message.slice(182, 186))
            });
        });

        discoverer.on('error', function (message, error) {
            if (timeoutId !== null) {
                clearTimeout(timeoutId);
            }
            debug(error);
            try {
                discoverer.close();
            } catch (ex) {/*ignore*/}
            reject(error);
        });
    });
};


demo.js
var smartplug = require('./index');
var options = {
    timeout: 10000,
    name:'edimax',
    host:'192.168.178.65',
    username: 'admin',
    password: '1234'
};


smartplug.getDeviceInfo(options).then(function (info) {
    console.log(info);
}).catch(function(e) {console.log("Request failed: ", e)});

smartplug.getSchedule(options).then(function (schedule) {
    console.log(schedule);
}).catch(function(e) {console.log("Request failed: ", e)});

// set switch ON
smartplug.setSwitchState(true, options).catch(function(e) {console.log("Request failed: ", e)});

smartplug.getSwitchPower(options).then(function (power) {
    console.log("Current switch power", power, "Watts");
}).catch(function(e) {console.log("Request failed: ", e)});

smartplug.getSwitchEnergy(options).then(function (energy) {
    console.log("getSwitchEnergy result:", energy);
}).catch(function(e) {console.log("Request failed: ", e)});

smartplug.getStatusValues(true, options).then(function (all) {
    console.log("getStatusValues result:", all);
}).catch(function(e) {console.log("Request failed: ", e)});

// set switch OFF
smartplug.setSwitchState(false, options).catch(function(e) {console.log("Request failed: ", e)});

// get switch status
smartplug.getSwitchState(options).then(function (state) {
    console.log("Switch status is", state?"ON":"OFF");
}).catch(function(e) {console.log("Request failed: ", e)});

// get schedule status
smartplug.getScheduleState(options).then(function (state) {
    console.log("Schedule status is", state?"ON":"OFF");
}).catch(function(e) {console.log("Request failed: ", e)});

// get the daily history of power measured consumption for the given date range
smartplug.getHistory('DAY', '20160825', '20160830', options).then(function (results) {
    console.log("getHistory result", results);
}).catch(function(e) {console.log("Request failed: ", e)});

// discover devices
smartplug.discoverDevices({
    timeout: 3000,
    address: "192.168.178.255"
}).then(function (results) {
    console.log("Discovery Result:", results);
}).catch(function(e) {console.log("Request failed: ", e)});
